### PR TITLE
feat(`campaign`): create min and max supply params

### LINKED ISSUE #493: Create params `MinSupply` and `MaxSupply`
We currently prevent when send `MsgUpdateTotalSupply` removing an existing denomination from the `TotalSupply` of the campaign otherwise the coordinator has the ability to remove value from already minted vouchers. The existing denom can only be set to a new value or new denoms can be added in the total supply

However, the value of the supply for a denom has no range and can be set to `1`. We should create params to set of reasonable range for any use case for the minimum value and the maximum value of a total supply

### PR DESCRIPTION
**Pull request by giunatale**
Closes #493. 

### What does this PR does?

Adds a `TotalSupplyRange` param to the `campaign` module. This param is a struct comprised of two `sdk.Int` values `MinTotalSupply` and `MaxTotalSupply` that define the range for a valid total supply.

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Include tests
- [x] Respect code style and lint

### CODE REVIEW

File: x/campaign/types/params_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		{
+			name:   "invalid range",
+			params: NewParams(DefaultMaxTotalSupply, DefaultMinTotalSupply),
+			err:    errors.New("maximum total supply should be greater than greater or equal than minimum total supply"),
+		},
+		{
+			name:   "valid range",
+			params: NewParams(DefaultMinTotalSupply, DefaultMaxTotalSupply),
+		},
+	}
```
Comment (aljo242) [inline x/campaign/types/params_test.go:27]:
```suggestion
		{
			name:   "invalid min total supply",
			params: NewParams(sdk.ZeroInt(), DefaultMaxTotalSupply),
		},
	
	}
```

File: x/campaign/types/total_supply.go [hunk 2]
```diff
@@ -21,3 +25,34 @@ func UpdateTotalSupply(coins, updatedCoins sdk.Coins) sdk.Coins {
 
 	return append(updatedCoins, notUpdated...).Sort()
 }
+
+// ValidateTotalSupply checks whether the total supply for each denom is within the provided total supply range
+func ValidateTotalSupply(coins sdk.Coins, supplyRange TotalSupplyRange) error {
```
Comment (aljo242) [inline x/campaign/types/total_supply.go:30]:
```suggestion
func ValidateCoinSupplies(coins sdk.Coins, supplyRange TotalSupplyRange) error {
```

I think this name is a bit clearer

**Review by aljo242 — CHANGES_REQUESTED** [review]

File: x/campaign/types/total_supply.go (hunk 2 above)
Comment (lubtd) [reply x/campaign/types/total_supply.go:30]: `TotalSupply` is our current term in the official terminology of SPN to designate the supply of tokens of a campaign's mainnet. If you have idea for renaming this should be further discussed in a call because there are many places where it would change
I believe Total Supply is a good term because it is very used in the ecosystem

File: proto/campaign/params.proto [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    (gogoproto.nullable)   = false
+  ];
+}
+
+// Params defines the set of params for the campaign module.
+message Params {
+  option (gogoproto.goproto_stringer) = false;
+
+  TotalSupplyRange totalSupplyRange = 1 [(gogoproto.nullable) = false];
+}
```
Comment (lubtd) [inline proto/campaign/params.proto:25]: We should put `Params` definition first in the file

File: testutil/sample/campaign.go [hunk 4]
```diff
@@ -77,10 +77,25 @@ func MainnetAccount(campaignID uint64, address string) campaign.MainnetAccount {
 
 // MsgCreateCampaign returns a sample MsgCreateCampaign
 func MsgCreateCampaign(coordAddr string) campaign.MsgCreateCampaign {
+	coins := Coins()
+	for i := range coins {
+		coins[i].Amount = coins[i].Amount.Add(campaign.DefaultMinTotalSupply)
+	}
```
Comment (lubtd) [inline testutil/sample/campaign.go:83]: I think we can introduce a `TotalSupply()` method in `sample` that initializes a `Coins` with token in the total supply range

File: x/campaign/keeper/keeper.go [hunk 5]
```diff
@@ -3,6 +3,8 @@ package keeper
 import (
 	"fmt"
 
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
```
Comment (lubtd) [inline x/campaign/keeper/keeper.go:6]: We should normalize the import order
```
"fmt"

"github.com/cosmos/cosmos-sdk/codec"
sdk "github.com/cosmos/cosmos-sdk/types"
paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
"github.com/tendermint/tendermint/libs/log"

"github.com/tendermint/spn/x/campaign/types"
```

File: x/campaign/types/genesis_test.go [hunk 6]
```diff
@@ -306,3 +309,43 @@ func TestGenesisState_Validate(t *testing.T) {
 		})
 	}
 }
+
+func TestGenesisState_ValidateParams(t *testing.T) {
```
Comment (lubtd) [inline x/campaign/types/genesis_test.go:313]: Params validaty test cases should be complete in the `types/params_test.go`. For genesis state we assume `params.ValdiateBasics()` is fully tested and therefore we should just test once that the genesis state is invalid if the params are invalid

File: x/campaign/types/params.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// ValidateBasic performs basic validation on campaign parameters.
+func (p Params) ValidateBasic() error {
+	if p.TotalSupplyRange.MinTotalSupply.LT(sdk.OneInt()) {
+		return errors.New("minimum total supply should be greater than one")
+	}
+	if p.TotalSupplyRange.MaxTotalSupply.LT(p.TotalSupplyRange.MinTotalSupply) {
+		return errors.New("maximum total supply should be greater than greater or equal than minimum total supply")
+	}
```
Comment (lubtd) [inline x/campaign/types/params.go:65]: Mmh I think we can simply use
```
return validateTotalSupplyRange(p.TotalSupplyRange)
```

File: x/campaign/types/params_test.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+			},
+			err: errors.New("parameter minTotalSupply cannot be less than one"),
+		},
+		{
+			name: "invalid range",
+			supplyRange: TotalSupplyRange{
+				MinTotalSupply: DefaultMaxTotalSupply,
+				MaxTotalSupply: DefaultMinTotalSupply,
+			},
+			err: errors.New("parameter maxTotalSupply cannot be less than minTotalSupply"),
```
Comment (lubtd) [inline x/campaign/types/params_test.go:66]: They're actually equal in this case, and I don't think we should prevent this situation, even though it shouldn't be set like this in practice, this means there is a single total supply value that can be set of all coin

File: x/campaign/types/total_supply.go (hunk 2 above)
Comment (lubtd) [inline x/campaign/types/total_supply.go:30]:
```suggestion
func ValidateTotalSupply(totalSupply sdk.Coins, supplyRange TotalSupplyRange) error {
```

File: x/campaign/types/total_supply.go [hunk 9]
```diff
@@ -21,3 +25,34 @@ func UpdateTotalSupply(coins, updatedCoins sdk.Coins) sdk.Coins {
...
+			supplyRange.MinTotalSupply.String(),
+			supplyRange.MaxTotalSupply.String(),
+		)
+	}
+
+	for _, coin := range coins {
+		totalSupply := coin.Amount
+		if totalSupply.LT(supplyRange.MinTotalSupply) {
+			return fmt.Errorf(
+				"total supply for %s cannot be less than minTotalSupply %s: %s",
```
Comment (lubtd) [inline x/campaign/types/total_supply.go:44]:
```suggestion
				"total supply for %s cannot be lower than minTotalSupply %s: %s",
```

File: x/campaign/types/total_supply_test.go [hunk 10]
```diff
@@ -60,3 +60,47 @@ func TestUpdateTotalSupply(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateTotalSupply(t *testing.T) {
+	newCoins := func(coinsStr string) sdk.Coins {
```
Comment (lubtd) [inline x/campaign/types/total_supply_test.go:65]: FYI, https://github.com/tendermint/spn/pull/520 will introduce a package `testutil/constructor` to centralized methods for this type of initialization

We can see depending on the first PR merged

File: x/campaign/types/total_supply.go [hunk 11]
```diff
@@ -21,3 +25,34 @@ func UpdateTotalSupply(coins, updatedCoins sdk.Coins) sdk.Coins {
 
 	return append(updatedCoins, notUpdated...).Sort()
 }
+
+// ValidateTotalSupply checks whether the total supply for each denom is within the provided total supply range
+func ValidateTotalSupply(coins sdk.Coins, supplyRange TotalSupplyRange) error {
+	// extra safety check, this should never happen
+	if supplyRange.MaxTotalSupply.LT(supplyRange.MinTotalSupply) {
```
Comment (lubtd) [inline x/campaign/types/total_supply.go:32]: I think it would make sense to introduce a `func (TotalSupplyRange) Validate() error` method so the check is always performed from a single entrypoint
Then the param validate calls this method

File: x/campaign/types/total_supply.go [hunk 12]
```diff
@@ -21,3 +25,34 @@ func UpdateTotalSupply(coins, updatedCoins sdk.Coins) sdk.Coins {
 
 	return append(updatedCoins, notUpdated...).Sort()
 }
+
+// ValidateTotalSupply checks whether the total supply for each denom is within the provided total supply range
+func ValidateTotalSupply(coins sdk.Coins, supplyRange TotalSupplyRange) error {
+	// extra safety check, this should never happen
```
Comment (lubtd) [inline x/campaign/types/total_supply.go:31]: I think it's actually a good practive to include these checks
In this case since it is an inconsistent state error, this should be handled by a `CriticalError`

Since the package `types` is stateless and unopinionated from `CriticalError`, this method should return a specific error type at this stage `ErrInvalidSupplyrange` then the calling methods in `keeper` should return themselves a `CriticalError` if `ErrInvalidSupplyrange` occurs

File: x/campaign/keeper/msg_server_create_campaign.go [hunk 13]
```diff
@@ -19,6 +19,12 @@ func (k msgServer) CreateCampaign(goCtx context.Context, msg *types.MsgCreateCam
 		return nil, sdkerrors.Wrap(profiletypes.ErrCoordAddressNotFound, msg.Coordinator)
 	}
 
+	// Validate provided totalSupply
+	totalSupplyRange := k.TotalSupplyRange(ctx)
+	if err := types.ValidateTotalSupply(msg.TotalSupply, totalSupplyRange); err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalidTotalSupply, err.Error())
+	}
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_create_campaign.go:26]: We must introduce a test case in message test for this condition

File: x/campaign/keeper/msg_server_update_total_supply.go [hunk 14]
```diff
@@ -35,6 +35,12 @@ func (k msgServer) UpdateTotalSupply(goCtx context.Context, msg *types.MsgUpdate
 		return nil, sdkerrors.Wrapf(types.ErrMainnetInitialized, "%d", msg.CampaignID)
 	}
 
+	// Validate provided totalSupply
+	totalSupplyRange := k.TotalSupplyRange(ctx)
+	if err := types.ValidateTotalSupply(msg.TotalSupplyUpdate, totalSupplyRange); err != nil {
+		return nil, sdkerrors.Wrap(types.ErrInvalidTotalSupply, err.Error())
+	}
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_update_total_supply.go:42]: We must introduce a test case in message test for this condition

**Review by lubtd — CHANGES_REQUESTED** [review]

File: x/campaign/types/total_supply.go (hunk 2 above)
Comment (giunatale) [reply x/campaign/types/total_supply.go:30]: Agreed with @lubtd, that was my intention when writing the function. I'd keep it as it is tbh

File: x/campaign/types/params_test.go (hunk 8 above)
Comment (giunatale) [reply x/campaign/types/params_test.go:66]: They aren't equal (`DefaultMinTotalSupply` used for max and `DefaultMaxTotalSupply` used for min hence max < min (?) ), and the situation in which the two range boundaries are equal is allowed.

File: x/campaign/types/genesis_test.go (hunk 6 above)
Comment (giunatale) [reply x/campaign/types/genesis_test.go:313]: I see `lanch/types/genesis_test.go` and I should just remove the last test case (the valid one) to make it consistent with `TestGenesisState_ValidateParams`. Is this what you are asking?

File: x/campaign/types/total_supply.go (hunk 9 above)
Comment (giunatale) [reply x/campaign/types/total_supply.go:44]: This is arbitrary, it's american vs british english imho (but I don't think we have chosen one or the other for style afaik), but fine I'll change it

File: x/campaign/keeper/msg_server_update_total_supply.go (hunk 14 above)
Comment (giunatale) [reply x/campaign/keeper/msg_server_update_total_supply.go:42]: right, I was thinking about it yesterday when I logged off, completely forgot to add test cases for this check in both messages. Will do

File: testutil/sample/campaign.go (hunk 4 above)
Comment (giunatale) [reply testutil/sample/campaign.go:83]: I' also adding a `CoinWithRange(min, max)` and `CoinsWithRange(min, max)` to `sample` since I need it for a test and perhaps having sample functions with boundaries for amounts might be useful for other things

File: x/campaign/types/params_test.go (hunk 1 above)
Comment (giunatale) [reply x/campaign/types/params_test.go:32]: added

File: x/campaign/keeper/msg_server_update_total_supply.go [hunk 15]
```diff
@@ -35,6 +37,15 @@ func (k msgServer) UpdateTotalSupply(goCtx context.Context, msg *types.MsgUpdate
 		return nil, sdkerrors.Wrapf(types.ErrMainnetInitialized, "%d", msg.CampaignID)
 	}
 
+	// Validate provided totalSupply
+	totalSupplyRange := k.TotalSupplyRange(ctx)
+	if err := types.ValidateTotalSupply(msg.TotalSupplyUpdate, totalSupplyRange); err != nil {
+		if errors.Is(err, types.ErrInvalidSupplyRange) {
```
Comment (giunatale) [inline x/campaign/keeper/msg_server_update_total_supply.go:43]: This branch here is not covered from tests (same in `msg_create_campaign`) since it would require to create a state where the range params are invalid. I assumed this would be fine

Comment (lubtd) [reply x/campaign/keeper/msg_server_update_total_supply.go:43]: Yes, critical errors are generally (except some exception) not tested simply because can't create a state where those are triggered unless we do some hacks in the store which is not in the idea of message tests

**Comment by lubtd** [comment]
@giunatale don't hesitate to resolve the conversions when you consider them as resolved

File: x/campaign/types/genesis_test.go (hunk 6 above)
Comment (lubtd) [reply x/campaign/types/genesis_test.go:313]: I'm thinking about two test cases:
- Genesis state is valid
- Genesis state is invalid because params are invalid (whatever the reason is)

File: x/campaign/keeper/msg_server_update_total_supply_test.go [hunk 16]
```diff
@@ -99,6 +99,15 @@ func TestMsgUpdateTotalSupply(t *testing.T) {
 			},
 			err: types.ErrMainnetInitialized,
 		},
+		{
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_update_total_supply_test.go:102]: Same comment as above

Maybe it can be worth for check all places where `TotalSupply` is assigned for test purposes and use `sample.TotalSupply()`

File: x/campaign/types/total_supply.go [hunk 17]
```diff
@@ -21,3 +26,42 @@ func UpdateTotalSupply(coins, updatedCoins sdk.Coins) sdk.Coins {
...
+				coin.Denom,
+				supplyRange.MaxTotalSupply.String(),
+				coin.Amount.String())
+		}
+	}
+	return nil
+}
+
+// Validate performs basic validation on an instance of TotalSupplyRange
+func (sr TotalSupplyRange) Validate() error {
```
Comment (lubtd) [inline x/campaign/types/total_supply.go:57]: We should also have `MinTotalSupply` being positive in the check

File: x/campaign/types/params_test.go (hunk 8 above)
Comment (lubtd) [reply x/campaign/types/params_test.go]: Sorry I misread

File: x/campaign/types/total_supply.go (hunk 17 above)
Comment (giunatale) [reply x/campaign/types/total_supply.go:57]: Since it became identical to `validateTotalSupplyRange()` I replaced it with a call to that function

Comment (lubtd) [reply x/campaign/types/total_supply.go:57]: There is maybe the interface conversion in addition in `validateTotalSupplyRange`

Comment (giunatale) [reply x/campaign/types/total_supply.go:57]: perhaps we could do the other way around? (use this function in `validateTotalSupplyRange`?)

[the remark this answers is not in the document]
Comment (giunatale) [reply x/campaign/types/total_supply.go:57]: To avoid duplication I ended up calling `TotalSupplyRange.ValidateBasic()` inside the param `validateTotalSupplyRange`). Thus, I've added test cases for  `TotalSupplyRange.ValidateBasic()` that cover all cases while in `validateTotalSupplyRange` I'm just testing: wrong interface, wrong param, valid param.

File: x/campaign/types/total_supply_test.go (hunk 10 above)
Comment (lubtd) [reply x/campaign/types/total_supply_test.go:65]: Can be used now :)

**Comment by lubtd** [comment]
Hey @giunatale, is it ready for review? Don't hesitate to push `Re-request review` button when it's the case ;)

File: x/campaign/types/genesis_test.go (hunk 6 above)
Comment (lubtd) [reply x/campaign/types/genesis_test.go:313]: Yes good point, we added more test case than necessary. I believe we can change this in https://github.com/tendermint/spn/issues/525 if we can consider this issue as a general refactoring of the `launch param`

File: x/campaign/types/total_supply_test.go (hunk 10 above)
Comment (giunatale) [reply x/campaign/types/total_supply_test.go]: rebased on `develop` and used `tc.Coins()` also here :)

**Review by lubtd — CHANGES_REQUESTED** [review]
Everything looks LGTM to me. We should just add a `Params` query as we have for other modules

**Comment by giunatale** [comment]
I missed that. makes sense

File: x/campaign/types/params_test.go [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+	tests := []struct {
+		name        string
+		supplyRange interface{}
+		err         error
+	}{
+		{
+			name:        "invalid interface",
+			supplyRange: "test",
+			err:         fmt.Errorf("invalid parameter type: string"),
+		},
```
Comment (Pantani) [inline x/campaign/types/params_test.go:56]: we should test the `MinTotalSupply` lower than one too

Comment (giunatale) [reply x/campaign/types/params_test.go:56]: we have decided to not fully test validation except where the core functionality is implemented. In all methods using that, we just have a single valid and invalid tests. @lubtd can confirm

File: x/campaign/types/genesis_test.go [hunk 19]
```diff
@@ -306,3 +309,36 @@ func TestGenesisState_Validate(t *testing.T) {
...
+		genState      types.GenesisState
+		shouldBeValid bool
+	}{
+		{
+			desc: "max total supply below min total supply",
+			genState: types.GenesisState{
+				Params: types.NewParams(types.DefaultMinTotalSupply, types.DefaultMinTotalSupply.Sub(sdk.OneInt())),
+			},
+			shouldBeValid: false,
+		},
```
[the remark this answers is not in the document]
Comment (giunatale) [reply x/campaign/types/genesis_test.go:325]: same as above

**Review by Pantani — APPROVED** [review]

**Review by lubtd — APPROVED** [review]
